### PR TITLE
[enhancement][Storage] refactor create predicate

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/olap/predicate_creator.h [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+class PredicateCreator {
+public:
+    virtual ColumnPredicate* create(const TabletColumn& column, int index,
+                                    const ConditionType& conditions, bool opposite,
+                                    MemPool* pool) = 0;
+    virtual ~PredicateCreator() = default;
+};
+
+template <typename CppType, PredicateType PT, typename ConditionType>
+class IntergerPredicateCreator : public PredicateCreator<ConditionType> {
```
Comment (Gabriel39) [inline be/src/olap/predicate_creator.h:40]:
```suggestion
class IntegerPredicateCreator : public PredicateCreator<ConditionType> {
```

File: be/src/olap/predicate_creator.h [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    return get_creator<PT, ConditionType>(column.type())
+            ->create(column, index, conditions, opposite, pool);
+}
+
+template <PredicateType PT>
+inline ColumnPredicate* create_comparison_predicate(const TabletColumn& column, int index,
+                                                    const std::string& condition, bool opposite,
+                                                    MemPool* pool) {
+    if (!PredicateTypeTraits::is_comparison(PT)) {
+        LOG(FATAL) << "input PredicateType is invalid, PredicateType=" << int(PT);
```
Comment (Gabriel39) [inline be/src/olap/predicate_creator.h:244]: Use `static_assert`

**Review by Gabriel39 — CHANGES_REQUESTED** [review]

File: be/src/olap/predicate_creator.h [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+};
+
+template <typename CppType, PredicateType PT, typename ConditionType>
+class IntergerPredicateCreator : public PredicateCreator<ConditionType> {
+public:
+    ColumnPredicate* create(const TabletColumn& column, int index, const ConditionType& conditions,
+                            bool opposite, MemPool* pool) override {
+        if constexpr (PredicateTypeTraits::is_list(PT)) {
+            phmap::flat_hash_set<CppType> values;
+            for (auto condition : conditions) {
```
Comment (HappenLee) [inline be/src/olap/predicate_creator.h:46]: const auto& condition

File: be/src/olap/predicate_creator.h [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+};
+
+template <typename CppType, PredicateType PT, typename ConditionType>
+class DecimalPredicateCreator : public PredicateCreator<ConditionType> {
+public:
+    ColumnPredicate* create(const TabletColumn& column, int index, const ConditionType& conditions,
+                            bool opposite, MemPool* pool) override {
+        if constexpr (PredicateTypeTraits::is_list(PT)) {
+            phmap::flat_hash_set<CppType> values;
+            for (auto condition : conditions) {
```
Comment (HappenLee) [inline be/src/olap/predicate_creator.h:71]: same to the up

File: be/src/olap/predicate_creator.h [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+                values.insert(convert(condition));
+            }
+            return new InListPredicateBase<CppType, PT>(index, std::move(values), opposite);
+        } else if constexpr (PredicateTypeTraits::is_comparison(PT)) {
+            return new ComparisonPredicateBase<CppType, PT>(index, convert(conditions), opposite);
+        }
+        return nullptr;
```
Comment (HappenLee) [inline be/src/olap/predicate_creator.h:53]: maybe need logfatal?

File: be/src/olap/predicate_creator.h [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+
+        return StringValue(buffer, length);
+    }
+};
+
+template <typename CppType, PredicateType PT, typename ConditionType>
+struct CustomPredicateCreator : public PredicateCreator<ConditionType> {
+public:
+    CustomPredicateCreator(std::function<CppType(const std::string& condition)> convert)
+            : _convert(convert) {};
```
Comment (HappenLee) [inline be/src/olap/predicate_creator.h:131]: pass the reference of `convert`？

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**